enterprise/server/backends/userdb/userdb.go

```diff
@@ -656,60 +671,73 @@ func (d *UserDB) addUserToGroup(ctx context.Context, tx *db.DB, userID, groupID
 		if existing.MembershipStatus == int32(grpb.GroupMembershipStatus_REQUESTED) {
 			return tx.Exec(`
 				UPDATE UserGroups
-				SET membership_status = ?
+				SET membership_status = ?, role = ?
 				WHERE user_user_id = ?
 				AND group_group_id = ?
 				`,
 				grpb.GroupMembershipStatus_MEMBER,
+				r,
 				userID,
 				groupID,
 			).Error
 		}
 		return status.AlreadyExistsError("You're already in this organization.")
 	}
-	row := &struct{ Count int64 }{}
-	err = tx.Raw(
-		"SELECT COUNT(*) AS count FROM UserGroups WHERE group_group_id = ?", groupID,
-	).Take(row).Error
-	if err != nil {
-		return err
-	}
-	r := role.Default
-	// If no existing users in the group, promote to admin automatically.
-	if row.Count == 0 {
-		r = role.Admin
-	}
 	return tx.Exec(
 		"INSERT INTO UserGroups (user_user_id, group_group_id, membership_status, role) VALUES(?, ?, ?, ?)",
 		userID, groupID, int32(grpb.GroupMembershipStatus_MEMBER), r,
 	).Error
 }
 
-func (d *UserDB) RequestToJoinGroup(ctx context.Context, userID string, groupID string) error {
+func (d *UserDB) RequestToJoinGroup(ctx context.Context, groupID string) (grpb.GroupMembershipStatus, error) {
+	u, err := perms.AuthenticatedUser(ctx, d.env)
+	if err != nil {
+		return 0, err
+	}
+	userID := u.GetUserID()
 	if userID == "" {
-		return status.InvalidArgumentError("User ID is required.")
+		return 0, status.InvalidArgumentError("User ID is required.")
 	}
 	if groupID == "" {
-		return status.InvalidArgumentError("Group ID is required.")
+		return 0, status.InvalidArgumentError("Group ID is required.")
 	}
-	return d.h.Transaction(ctx, func(tx *db.DB) error {
-		existing, err := getUserGroup(tx, userID, groupID)
+	var newStatus grpb.GroupMembershipStatus
+	err = d.h.Transaction(ctx, func(tx *db.DB) error {
+		tu, err := d.getUser(tx, u.GetUserID())
 		if err != nil {
 			return err
 		}
-		if existing == nil {
-			return tx.Create(&tables.UserGroup{
-				UserUserID:       userID,
-				GroupGroupID:     groupID,
-				Role:             uint32(role.Default),
-				MembershipStatus: int32(grpb.GroupMembershipStatus_REQUESTED),
-			}).Error
+		group, err := d.getGroupByID(tx, groupID)
+		if err != nil {
+			return err
 		}
-		if existing.MembershipStatus == int32(grpb.GroupMembershipStatus_REQUESTED) {
+		// If the org has an owned domain that matches the user's email,
+		// the user can join directly as a member.
+		newStatus = grpb.GroupMembershipStatus_REQUESTED
+		if group.OwnedDomain != "" && group.OwnedDomain == getEmailDomain(tu.Email) {
+			newStatus = grpb.GroupMembershipStatus_MEMBER
+			return d.addUserToGroup(tx, userID, groupID)
+		}
+
+		// Check if there's an existing request and return AlreadyExists if so.
+		existing, err := getUserGroup(tx, userID, groupID)
+		if err != nil {
+			return err
+		}
+		if existing != nil {
 			return status.AlreadyExistsError("You've already requested to join this organization.")
 		}
```

Comment:
do we want to distinguish between already requested and already a member?

Reply:
Woops, missed this comment. We do distinguish these 2 cases - the "already a member" case is just moved into the `addUserToGroup` func

Reply:
Actually nvm - that addUserToGroup is only called in the owned-domain case. Will fix and add a test

Reply:
Done - PTAL